### ISSUE TITLE
Hardening: bcrypt cost and not hardcoding which client headers to trust

### ISSUE THREAD

**Issue by chrissawyerfan4** [Kovah/LinkAce#617]
Sorry for mixing issues but it seemed a bit silly to open separate tickets for each one.

1. The hash created in the database is bcrypt with a cost setting of 10. This was good advice in 2013, but while Moore's (loosely interpreted) law is not really holding up anymore, this is a little bit dated. My hardware is from the early 2010s and even I set it to 12 (note that it is quadratic: 12 is two doublings above 10). The common scenario is probably not to log in on a daily basis, so most users would probably be happy with an even higher setting as well.

2. [The code](https://github.com/Kovah/LinkAce/blob/main/app/Http/Middleware/TrustProxies.php) seems to just accept basically any header as the truth. Without further auditing, I don't know what this is used for, but since it is used, presumably there is a purpose to knowing the IP address of the user. This code allows an attacker to choose their apparent IP address and is also not very flexible due to hardcoding. Depending on how this is used (or may be used in the future), it would be safer to default to trusting only the non-user-input `REMOTE_ADDR` and letting the user configure a proxy header, if any, in the `.env` file or similar.

3. Sneaking in one more concern, I tried toggling [`\Fruitcake\Cors\HandleCors`](https://github.com/Kovah/LinkAce/blob/main/app/Http/Kernel.php#L18) middleware but do not see the difference in HTTP response headers. Is this package used currently? Particularly for the application (not the API), I don't think any CORS headers should be set anyway, but it's in the global middleware stack rather than in the API-specific one. Removing it did not lead to any issues on my side (but I don't use the API). It was also deprecated two weeks ago, I just noticed.

---

For the first issue, I'm not sure where this should be configured, so I simply did / one workaround is to use this command and set the password field to the resultant value:

```bash
pwd="$(</dev/urandom tr -dc 0-9a-zA-Z | head -c 14)"
echo "$pwd"
time php -r 'echo password_hash("'"$pwd"'", PASSWORD_BCRYPT, ["cost"=>12]);'
```

Where the cost can be increased until the `time` comes back as being 250ms or more on the target system.

The other two middlewares I simply removed.

### LinkAce version

github

### Setup Method

PHP

### Operating System

Linux (Ubuntu, CentOS,...)

**Comment by Kovah** [comment]
Thanks for the report. I have updated the default values for bcrypt and argon. Additionally, i made the hashing driver configurable so that people could enable Argon if they can. The Bcrypt rounds can be configured too with BCRYPT_ROUNDS.

The proxy setting can be configured in the env file now. I have left out the header setting for now as it expects pretty specific values for different headers that are not easily accessible.

HandleCors was removed. It's a left-over from an older version of the framework.
